pkg/apitype/policy.go

```diff
@@ -33,6 +33,9 @@ type CreatePolicyPackRequest struct {
 	// The Policies outline the specific Policies in the package, and are derived
 	// from the package by the CLI.
 	Policies []Policy `json:"policies"`
+
+	// The JSON schema for the Policy Pack's configuration.
+	ConfigSchema map[string]interface{} `json:"configSchema,omitempty"`
```

Comment:
The `map[string]interface{}` "isn't great" IMHO, because we lose any type information. But we do this already in some of the engine event types for the same reason. And there isn't really a better alternative, so definitely 👍 .

JFYI, in another situation where we pass unknown data between the API and client, we send it as `json.RawMessage`. (See `apitype.VersionedCheckpoint` and callers.) But personally the `map[string]interface{}` is much more convenient and possibly more reliable.

Reply:
I have some thoughts/questions here... Will send feedback in a bit.

Reply:
Consider the following policies in a Policy Pack:

```typescript
{
    name: "efs-encrypted",
    description: "Checks whether EFS is configured to encrypt the file data using KMS.",
    config: {
        properties: {
            kmsKeyId: {
                type: "string",
                description: "ARN of the KMS key used to encrypt the EFS file system.",
            },
        },
    },
    validateResource: ...
},
{
    name: "another-policy-name",
    description: "Checks some important things.",
    config: {
        properties: {
            someRequiredProperty: {
                type: "number",
                description: "Some required property.",
            },
        },
        required: ["someRequiredProperty"],
    },
    validateResource: ...
}
```

Each policy in the policy pack has a set of config properties, and each of those config properties has a JSON schema.

Do we send the config schema up as part of each policy's metadata? e.g. add a `ConfigSchema` property to the `apitype.Policy`  struct, something like:

```go
type Policy struct {
	...

	// The JSON schema for the Policy (map of config property name to schema).
	ConfigSchema map[string]PolicyConfigSchema `json:"configSchema,omitempty"`
}


type PolicyConfigSchema struct {
	// Config property name to JSON Schema map.
	Properties: map[string]interface{} `json:"properties,omitempty"`
	// Required config properties.
	Required:   []string               `json:"required,omitempty"`
}
```

Or should the client combine the schemas of all config properties of all policies together into something like the following?

```json
"configSchema": {
    "efs-encrypted": {
        "properties": {
            "kmsKeyId": {
                "type": "string",
                "description": "ARN of the KMS key used to encrypt the EFS file system."
            }
        }
    },
    "another-policy-name": {
        "properties": {
            "someRequiredProperty": {
                "type": "number",
                "description": "Some required property."
            }
        },
        "required": ["someRequiredProperty"]
    }
}
```

In which case, it could be typed as:

```go
type CreatePolicyPackRequest struct {
	...

	// Map of policy name to config schema for the policy.
	ConfigSchema map[string]PolicyConfigSchema `json:"configSchema,omitempty"`
}
```

Or, does the client synthesize a JSON schema for the whole Policy Pack? (Treating the top-level Policy Pack’s schema as an object where each property is a policy name which is an object with each property being the config property.

```json
"configSchema": {
    "type": "object",
    "properties": {
        "efs-encrypted": {
            "type": "object",
            "properties": {
                "kmsKeyId": {
                    "type": "string",
                    "description": "ARN of the KMS key used to encrypt the EFS file system."
                }
            }
        },
        "another-policy-name": {
            "type": "object",
            "properties": {
                "someRequiredProperty": {
                    "type": "number",
                    "description": "Some required property."
                }
            },
            "required": ["someRequiredProperty"]
        }
    }
}
```

In which case, it could be typed as `map[string]interface{}` or `interface{}` or `json.RawMessage`:

```go
type CreatePolicyPackRequest struct {
 	// The JSON schema for the Policy Pack's configuration.
 	ConfigSchema map[string]interface{} `json:"configSchema,omitempty"`
}
```

Finally, if we are synthesizing a JSON schema for the whole policy pack, and are allowing the value of each policy name to be either a string `enforcementLevel` or `object` (per my comment [here](https://github.com/pulumi/pulumi/pull/3872/files#r375964463)), do we further synthesize the JSON schema to include that?

```json
"configSchema": {
    "definitions": {
        "enhancementLevel": {
            "type": "string",
            "enum": ["disabled", "advisory", "mandatory"]
        }
    },

    "type": "object",
    "properties": {
        "efs-encrypted": {
            "oneOf": [
                { "$ref": "#/definitions/enhancementLevel" },
                {
                    "type": "object",
                    "properties": {
                        "enhancementLevel": { "$ref": "#/definitions/enhancementLevel" },
                        "kmsKeyId": {
                            "type": "string",
                            "description": "ARN of the KMS key used to encrypt the EFS file system."
                        }
                    }
                }
            ]
        },
        "another-policy-name": {
            "oneOf": [
                { "$ref": "#/definitions/enhancementLevel" },
                {
                    "type": "object",
                    "properties": {
                        "enhancementLevel": { "$ref": "#/definitions/enhancementLevel" },
                        "someConfigurableProperty": {
                            "type": "number",
                            "description": "Some configurable property."
                        }
                    },
                    "required": ["someConfigurableProperty"]
                }
            ]
        }
    }
}
```

What we decide impacts the DB schema as well.

Reply:
> does the client synthesize a JSON schema for the whole Policy Pack?

This was what I initially had in mind -- though I can see an argument for sending the configuration alongside the `apitype.Policy`. Since we're planning on storing all the "enabled" configuration as one set of configuration, I think it makes most sense to synthesize one JSON schema for the pack.

I do think we should augment it with things like `enforcementLevel` and `all`.

Reply:
If we are going to do something like: 
```
type PolicyConfigSchema struct {
	// Config property name to JSON Schema map.
	Properties: map[string]interface{} `json:"properties,omitempty"`
	// Required config properties.
	Required:   []string               `json:"required,omitempty"`
}
```

we can slowly add support for other validation etc -- though I want to make sure whatever we decided we can still translate to a JSON schema and use an existing validation library.

---

I am trying to find the OPA documentation on their "parameters" to see how that could effect this.

Reply:
Taking a look at Azure Policies...

[Azure Policy example](https://github.com/Azure/azure-policy/blob/master/built-in-policies/policyDefinitions/App%20Service/AppService_ApiApp_Audit_PHP_Latest.json#L11)

[Azure PolicySet example](https://github.com/Azure/azure-policy/blob/master/built-in-policies/policySetDefinitions/Guest%20Configuration/GuestConfiguration_AzureBaseline.json#L11)

Looks like most parameters have this format:
```
            "AuditDetailedFileShare": {
                "type": "string",
                "metadata": {
                    "displayName": "Audit Detailed File Share",
                    "description": "If this policy setting is enabled, access to all shared files and folders on the system is audited. Auditing for Success can lead to very high volumes of events."
                },
                "defaultValue": "No Auditing",
                "allowedValues": [
                    "No Auditing",
                    "Success",
                    "Failure",
                    "Success and Failure"
                ]
            },
```

A Policy Set then specifies a Policy and [passes it parameters](https://github.com/Azure/azure-policy/blob/master/built-in-policies/policySetDefinitions/Guest%20Configuration/GuestConfiguration_AzureBaseline.json#L966-L978)

Reply:
The more I've been thinking about this, the more hesitant I am about synthesizing a JSON Schema for the entire policy pack. The two things I'm concerned about:

1. What happens if we allow [definitions/references](https://json-schema.org/understanding-json-schema/structuring.html) in the JSON schema, and multiple config properties define definitions that conflict with each other? How would we handle synthesizing these conflicting schemas into a single JSON schema?

2. When it comes time to leveraging the JSON schema for each config property in the front-end UI, it will be much simpler to have the schema for each individual config property, because then we can easily see what type of form element it should be (e.g. `<input type="number">` for `"type": "number"`, or array builder for `"type": "array"`, etc.) and run a validator against just the property value.

So, I'm leaning towards something like:

```go
type Policy struct {
	...

	// The JSON schema for the Policy (map of config property name to schema).
	ConfigSchema *PolicyConfigSchema `json:"configSchema,omitempty"`
}

type PolicyConfigSchema struct {
	// Config property name to JSON Schema map.
	Properties: map[string]*json.RawMessage `json:"properties,omitempty"`
	// Required config properties.
	Required:   []string                   `json:"required,omitempty"`
}
```

Reply:
My concern here would be that the storage of the configuration would become a lot more complex. Looking back at the [db schema PR](https://github.com/pulumi/pulumi-service/pull/4611/files), we then add the config_schema to the `Policies` table. Today we only use this table to display the Policies in the pack. 

A few things we would need to change:
* We would then need to query this table to do the schema validation on the front end.
* The `PolicyPackConfigs` would also need a `policy_name` column to reference the correct policy. (The `Policies` table PK is (`policy_pack_id`, `name`))
* For the `PolicyGroupPolicyPacks` and `ProgramUpdatePolicyPacks` things will become a lot more complex, since we will need to reference multiple `PolicyPackConfigs`
   * For both situations, we would likely need a second table to hold all the ids
   * Maybe the other option is to make a giant JSON document thats a map of `policy_name` to the configuration. In this case we could store each config_schema in the `Policies` table and then the configuration can still be stored in one entry of the `PolicyPackConfigs`. ---> But then in this case why couldnt we do something similar with the schema, where the schema isnt a proper JSON schema but instead a map of `policy_name` to JSON schema to use for that policy. This should allow you to have the same reference multiple times. 


```go
type CreatePolicyPackRequest struct {
	...
	// The JSON schema for the Policy Pack's configuration.
	ConfigSchemas map[string]*json.RawMessage `json:"configSchemas,omitempty"`
}
type PolicyConfigSchema struct {
	// Config property name to JSON Schema map.
	Properties: map[string]*json.RawMessage `json:"properties,omitempty"`
	// Required config properties.
	Required:   []string                   `json:"required,omitempty"`
}
```

Reply:
I don't think what I'm suggesting necessarily has to change the db schema as defined in the [db schema PR](https://github.com/pulumi/pulumi-service/pull/4611/files) at all.

I'm suggesting the API types look as follows:

We already have these structs:

```go
type CreatePolicyPackRequest struct {
    // Name is a unique URL-safe identifier (at the org level) for the package.
    // If the name has already been used by the organization, then the request will
    // create a new version of the Policy Pack (incremented by one). This is supplied
    // by the CLI.
    Name string `json:"name"`

    // A pretty name for the Policy Pack that is supplied by the package.
    DisplayName string `json:"displayName"`

    // The Policies outline the specific Policies in the package, and are derived
    // from the package by the CLI.
    Policies []Policy `json:"policies"`
}

// Policy defines the metadata for an individual Policy within a Policy Pack.
type Policy struct {
    // Unique URL-safe name for the policy.  This is unique to a specific version
    // of a Policy Pack.
    Name        string `json:"name"`
    DisplayName string `json:"displayName"`

    // Description is used to provide more context about the purpose of the policy.
    Description      string           `json:"description"`
    EnforcementLevel EnforcementLevel `json:"enforcementLevel"`

    // Message is the message that will be displayed to end users when they violate
    // this policy.
    Message string `json:"message"`
}
```

It seems very natural to include additional policy metadata on the `Policy` struct. I'm proposing we add an optional `ConfigSchema` property to `Policy` typed as a new `PolicyConfigSchema` struct:

```go
type Policy struct {
    ...

    // Optional config schema for the policy.
    ConfigSchema *PolicyConfigSchema `json:"configSchema,omitempty"`
}

type PolicyConfigSchema struct {
    // Config property name to JSON Schema map.
    Properties: map[string]*json.RawMessage `json:"properties,omitempty"`
    // Required config properties.
    Required:   []string                   `json:"required,omitempty"`
}
```

The `Properties` property is a map of config property names to their associated JSON schema.

The `Required` property is an array of required properties.

Now, in terms of the storage on the service side, if we want to store config schema information for the entire policy pack in a single column as a JSON blob, we still can. I'd suggest the service use a _simple_ JSON format similar to the following, rather than trying to synthesize an "uber JSON schema" for the entire policy pack. The service would just create a JSON document the looked as follows and store it as a blob in the db:

```JSON
{
    "a-policy-name": {
        "properties": {
            "kmsKeyId": {
                "type": "string",
                "description": "ARN of the KMS key used to encrypt the EFS file system."
            }
        }
    },
    "another-policy-name": {
        "properties": {
            "someRequiredProperty": {
                "type": "number",
                "description": "Some required property."
            }
        },
        "required": ["someRequiredProperty"]
    }
}
```

This JSON can be returned to the front-end, and it'd be very easy for the front-end to drill into each policy's config properties.

Though, we may want to structure the response to the front-end a bit more nicely in the `apitype`, making use of the proposed `PolicyConfigSchema` struct, something like:

```go
 // GetPolicyPackConfigSchemaResponse is the response that includes the JSON
 // of a particular Policy Pack's configuration.
 type GetPolicyPackConfigSchemaResponse struct {
 	// Map of policy name to config schema.
 	ConfigSchema: map[string] PolicyConfigSchema `json:"configSchema,omitempty"`
 }
```

Of course, if we wanted to add more structure in the DB we could consider adding something like a `PolicyConfigPropertiesSchema` table that references the `Policies` table, but I don't think we need to bother doing that right now as it seems more ceremony than it's worth.

Reply:
I'm sort of confused by the mix of explicit `properties` and `required` fields with the implicit map used for `properties`... because a property could have an object that would also have a `PolicyConfigSchema`? I think we should either be explicit or not, my preference being to be explicit and limit the scope of fields we support early on... thought? Or maybe i missing something and this is totally fine 🤔

Reply:
Let’s go over it tomorrow in person. The top-level `PolicyConfigSchema` is used to explicitly specify the set of config properties for the policy. It itself is not a JSON Schema (although the `properties` and `required` names are inspired by those found in JSON Schema). The `properties` map is a map of the config property name to its JSON Schema. The map value is a JSON Schema for the property. The value conforms to the JSON Schema standard (and is unrelated to and would not reuse `PolicyConfigSchema`). Any valid JSON schema should work for the value, though, we can limit what we allow folks to specify initially in the object model we expose in each language SDK. However, any valid JSON Schema should work, IMO.